### PR TITLE
Update android gradle plugin to 3.0.0

### PR DESCRIPTION
**Pull request by zaki50**
https://developer.android.com/studio/preview/features/new-android-plugin.html
https://developer.android.com/studio/preview/features/new-android-plugin-migration.html

When you switch android gradle plugin from 2.x to 3.x, you need to stop running gradle daemon by executing `./gradlew --stop`.

### CODE REVIEW

**Comment by cmelchior** [comment]
It also appears that `annotationProcessor` will automatically delegate to `kapt` if needed. That will reduce some complexity which is nice. We also need to completely remove `apt` since that won't work at all.

**Comment by zaki50** [comment]
@cmelchior nice!

**Comment by zaki50** [comment]
@cmelchior You mean we would drop the support for older Android Cradle Plugin?

**Comment by cmelchior** [comment]
I guess we need to figure out how to go about it. If we could handle both versions inside the same plugin that would obviously be preferable at least during a transition period, but I'm not sure how we can detect the difference. I assume we will have to rely on reflection

**Comment by zaki50** [comment]
I talked with Xav at the Office Hour in Google I/O and he said that he will add a feature to solve this compatibility issue.

**Comment by cmelchior** [comment]
Woop woop

**Comment by Zhuinden** [comment]
That's nice of them

**Comment by zaki50** [comment]
Now realm plugin supports both old and new android gradle plugins.

**Comment by zaki50** [comment]
https://github.com/robolectric/robolectric/issues/3169 seems to cause the test failure.

**Comment by zaki50** [comment]
I confirmed that disabling aapt2 solved the failure, but I don't disable it for a while to wait the fix of upstream projects.

**Comment by zaki50** [comment]
Disabled `aapt2` in example project to start merge process of this PR.

@realm/java Could you review this?

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 1]
```diff
@@ -40,7 +41,9 @@ class Realm implements Plugin<Project> {
         }
 
         def syncEnabledDefault = false
-        project.extensions.create('realm', RealmPluginExtension, project, syncEnabledDefault)
+        def dependencyConfigurationName = getDependencyConfigurationName(project)
+        project.extensions.create('realm', RealmPluginExtension,
```
Comment (cmelchior) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Couldn't this fit into one line?

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 2]
```diff
@@ -80,6 +83,17 @@ class Realm implements Plugin<Project> {
         }
     }
 
+    private static String getDependencyConfigurationName(Project project) {
```
Comment (cmelchior) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Perhaps add a comment here that `newDependencyName` is from `3.0.0-*` ?

File: realm-transformer/src/main/groovy/io/realm/transformer/RealmTransformer.groovy [hunk 3]
```diff
@@ -66,6 +66,7 @@ class RealmTransformer extends Transform {
 
     @Override
     Set<Scope> getReferencedScopes() {
+        // Scope.PROJECT_LOCAL_DEPS and Scope.SUB_PROJECTS_LOCAL_DEPS is only for the compatibility with AGP 1.x, 2.x
```
Comment (cmelchior) [inline realm-transformer/src/main/groovy/io/realm/transformer/RealmTransformer.groovy]: only for the -> only for

**Comment by zaki50** [comment]
@cmelchior I tested ithis with old plugin  before, but will test it again to make sure.

**Comment by beeender** [comment]
I got `Error:This Gradle plugin requires Studio 3.0 minimum` on AS 2.3.3 . Is it OK for everyone to update the AS? @realm/java I am OK to update it :P

**Comment by zaki50** [comment]
@beeender Ah, good catch!

Recently I always use AS 3.0 Canary and didn't see any major problems. @realm/java

**Comment by cmelchior** [comment]
I'm already using 3.0.0-alpha4 so... :)

**Comment by zaki50** [comment]
@cmelchior I tested the local build of this branch with Java and Kotlin projects that were using android gradle plugin 2.3.3. Worked fine!

**Pull request merged**